graalpy/graalpy-scripts-debug-guide/README.md

```diff
@@ -0,0 +1,321 @@
+                target.setText(String.format("%s = %f x %s", file0.getName(), result, file1.getName()));
+                success = true;
+            } catch (RuntimeException e) {
+                target.setText(e.getMessage());
+            }
+        }
+        resetTargetColor(target);
+        event.setDropCompleted(success);
+        event.consume();
+    });
+}
+```
+
+❶ We can pass a resource URL to the GraalVM Polyglot [`Source`](https://docs.oracle.com/en/graalvm/enterprise/20/sdk/org/graalvm/polyglot/Source.html) API.
+The content is read by the `Source` object, GraalPy and the Python code do not gain access to Java resources this way.
+
+❷ Python objects are returned using a generic [`Value`](https://docs.oracle.com/en/graalvm/enterprise/20/sdk/org/graalvm/polyglot/Value.html) type.
+
+❸ As a Python function, `compare_files` can be executed.
+GraalPy accepts Java objects and tries to match them to the appropriate Python types.
+Return values are again represented as `Value`.
+In this case we know the result will be a Python `float`, which can be converted to a Java `double`.
+
+## 5. Running the application
+
+If you followed along with the example, you can now compile and run your application from the commandline:
+
+With Maven:
+
+```shell
+./mvnw compile
+./mvnw javafx:run
+```
+
+With Gradle:
+
+```shell
+./gradlew assemble
+./gradlwe run
+```
+
+**Note:** If you do not use the Maven or Gradle wrappers, the commands would just be `mvn` and `gradle` to use them from your `PATH`.
+
+## 5.1 Debugging embedded Python code
+
+Your Python code may behave differently when run in a Java embedding.
+This can have many reasons, from different types passed in from Java, permissions of the GraalVM Polyglot sandbox, to Python libraries assuming OS-specific process properties that Java applications do not expose.
+
+To debug Python scripts, we recommend you use VSCode.
+Make sure you have installed the [Python extensions](https://marketplace.visualstudio.com/items?itemName=ms-python.python).
+Where we build the Python context, we already added the following options to enable remote debugging:
+
+`App.java`
+```java
+.option("dap", "localhost:4711")
+.option("dap.Suspend", "false")
+```
+
+This instructs the runtime to accept [DAP]() connections on port 4711 and continue execution.
+We add a debug configuration to VSCode to match:
+
+`.vscode/launch.json`
+```json
+{
+    "configurations": [{
+        "name": "GraalPy: Attach embedded",
+        "type": "debugpy",
+        "request": "attach",
+        "connect": { "host": "localhost", "port": 4711 },
+    }]
+}
+```
+
+When we run the application, we see the following output:
+
+```
+[Graal DAP] Starting server and listening on localhost/127.0.0.1:4711
+```
+
+We can connect using VSCode or any other DAP client.
+The loaded sources can be opened to view the Python code as loaded from the Java resources.
+We can set breakpoints and inspect runtime state as we would expect.
```

Comment:
Maybe this can stress out that you should open the file from the debugger panel -> loaded sources, and not from explorer. It's visible in the gif, but at first my subconscious let to me open the file from explorer and put a breakpoint there.